src/keys/mod.rs

```diff
@@ -310,45 +373,153 @@ pub trait ToDescriptorKey<Ctx: ScriptContext>: Sized {
+///     }
+/// }
+/// ```
+///
+/// Types that don't internally encode the [`Network`](bitcoin::Network) in which they are valid need some extra
+/// steps to override the set of valid networks, otherwise only the network specified in the
+/// [`ExtendedPrivKey`] or [`ExtendedPubKey`] will be considered valid.
+///
+/// ```
+/// use bdk::bitcoin;
+/// use bdk::bitcoin::util::bip32;
+/// use bdk::keys::{
+///     any_network, DerivableKey, DescriptorKey, ExtendedKey, KeyError, ScriptContext,
+/// };
+///
+/// struct MyCustomKeyType {
+///     key_data: bitcoin::PrivateKey,
+///     chain_code: Vec<u8>,
+/// }
+///
+/// impl<Ctx: ScriptContext> DerivableKey<Ctx> for MyCustomKeyType {
+///     fn into_extended_key(self) -> Result<ExtendedKey<Ctx>, KeyError> {
+///         let xprv = bip32::ExtendedPrivKey {
+///             network: bitcoin::Network::Bitcoin, // pick an arbitrary network here
+///             depth: 0,
+///             parent_fingerprint: bip32::Fingerprint::default(),
+///             private_key: self.key_data,
+///             chain_code: bip32::ChainCode::from(self.chain_code.as_ref()),
+///             child_number: bip32::ChildNumber::Normal { index: 0 },
+///         };
+///
+///         xprv.into_extended_key()
+///     }
+///
+///     fn into_descriptor_key(
+///         self,
+///         source: Option<bip32::KeySource>,
+///         derivation_path: bip32::DerivationPath,
+///     ) -> Result<DescriptorKey<Ctx>, KeyError> {
+///         let descriptor_key = self
+///             .into_extended_key()?
+///             .into_descriptor_key(source, derivation_path)?;
+///
+///         // Override the set of valid networks here
+///         Ok(descriptor_key.override_valid_networks(any_network()))
+///     }
+/// }
+/// ```
+///
 /// [`DerivationPath`]: (bip32::DerivationPath)
-pub trait DerivableKey<Ctx: ScriptContext> {
-    /// Add a extra metadata, consume `self` and turn it into a [`DescriptorKey`]
-    fn add_metadata(
+/// [`ExtendedPrivKey`]: (bip32::ExtendedPrivKey)
+/// [`ExtendedPubKey`]: (bip32::ExtendedPubKey)
+pub trait DerivableKey<Ctx: ScriptContext = miniscript::Legacy>: Sized {
+    /// Consume `self` and turn it into an [`ExtendedKey`]
+    ///
+    /// This can be used to get direct access to `xprv`s and `xpub`s for types that implement this trait,
+    /// like [`Mnemonic`](bip39::Mnemonic) when the `keys-bip39` feature is enabled.
+    #[cfg_attr(
+        feature = "keys-bip39",
+        doc = r##"
+    use bdk::bitcoin::Network;
+    use bdk::keys::{DerivableKey, ExtendedKey};
+    use bdk::keys::bip39::{Mnemonic, Language};
+
+    # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let xkey: ExtendedKey =
+        Mnemonic::from_phrase(
+            "jelly crash boy whisper mouse ecology tuna soccer memory million news short",
+            Language::English
+        )?
+        .into_extended_key()?;
+    let xprv = xkey.into_xprv(Network::Bitcoin).unwrap();
+    # Ok(()) }
```

Comment:
need to add ``` between line 476 and 477

Reply:
I was getting errors when adding that so I assumed they were being added automatically. I tried building the docs locally and the code snippet was looking good.

Reply:
With `1.48.0` too chain this test also passed for me. But I was only able to get the doc tests to work with `1.49.0` if I added the first ``` on the same line as the opening r##" and then didn't seem to matter where i put the closing ``` as long as it was before the closing "##

Can you verify your default toolchain is stable `1.49.0` ?

Reply:
Ah yeah, it looks like I was also using `1.48` locally

Reply:
Getting these errors on `1.49.0`:

```
failures:

---- src/keys/mod.rs - keys::DerivableKey::into_extended_key (line 461) stdout ----
error: unknown start of token: `
  --> src/keys/mod.rs:475:5
   |
16 |     ```
   |     ^
   |
help: Unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it is not
   |
16 |     '``
   |     ^

error: unknown start of token: `
  --> src/keys/mod.rs:475:6
   |
16 |     ```
   |      ^
   |
help: Unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it is not
   |
16 |     `'`
   |      ^

error: unknown start of token: `
  --> src/keys/mod.rs:475:7
   |
16 |     ```
   |       ^
   |
help: Unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it is not
   |
16 |     ``'
   |       ^

error: aborting due to 3 previous errors

Couldn't compile the test.

failures:
    src/keys/mod.rs - keys::DerivableKey::into_extended_key (line 461)
```

Reply:
Ah okay, the solution is to remove the tab before every line. I think it considers it code if there's a tab before the line